pkg/server/controller.go

```diff
@@ -303,13 +303,15 @@ func (ctrl *Controller) serverMux() (http.Handler, error) {
 
 	// FIXME: Honor ctrl.config.RemoteRead.Enabled.
 	appsRouter := apiRouter.PathPrefix("/").Subrouter()
-	appsRouter.Use(
+	appsRouter.HandleFunc("/apps", ctrl.getAppsHandler()).Methods("GET")
+
+	protectedAppsRouter := apiRouter.PathPrefix("/").Subrouter()
+	protectedAppsRouter.Use(
 		api.AuthMiddleware(nil, ctrl.authService, ctrl.httpUtils),
 		authz.NewAuthorizer(ctrl.log, httputils.NewDefaultHelper(ctrl.log)).RequireOneOf(
 			authz.Role(model.AdminRole),
 		))
-	appsRouter.HandleFunc("/apps", ctrl.getAppsHandler()).Methods("GET")
-	appsRouter.HandleFunc("/apps", ctrl.deleteAppsHandler()).Methods("DELETE")
+	protectedAppsRouter.HandleFunc("/apps", ctrl.deleteAppsHandler()).Methods("DELETE")
```

Comment:
@kolesnikovae @korniltsev @petethepig suggestions on how to handle this in a more scalable manner?

Reply:
I would move this to line 220:
```go
	authorizer := authz.NewAuthorizer(ctrl.log, httputils.NewDefaultHelper(ctrl.log))
	requireAdminRole := authorizer.Require(authz.Role(model.AdminRole))

	appsRouter := apiRouter.PathPrefix("/apps").Subrouter()
	appsRouter.Path("/").
		Methods(http.MethodGet).HandlerFunc(ctrl.getAppsHandler()).
		Methods(http.MethodDelete).Handler(requireAdminRole(ctrl.deleteAppsHandler()))
```

Definitely out of scope for this PR but:
 - I would love to refactor the handler and service (interface only) to `/api` package
 - Noticed that `authz.NewAuthorizer(ctrl.log, httputils.NewDefaultHelper(ctrl.log))` takes logger twice but never uses the first one. Must be a leftover

Reply:
> Ugh this is confusing. 
>
> That being said still using requireAdminRole, which (I think) means even in a public instance it's non possible to delete an app, which is fair but confusing when running locally.

That's true. Currently, if authentication is disabled, no admin actions can be performed, corresponding UI components are not shown. Now that we way want to change this, we can make an option telling what permissions to grant to an anonymous user when authentication is not enabled.

Also, not directly related but we may also consider adding `Member` role which allows more actions, for example: adding/deleting annotations, uploading adhoc profiles and so on.

> Speaking of requireAdminRole, I saw that there's a https://github.com/pyroscope-io/pyroscope/blob/main/pkg/api/authz/authz.go#L24-L26 but its use seems to be used as a middleware, I guess we should unify that.

In some cases it is more convenient to use it as a middleware (the code you reference), in others - wrap a specific handler. Typically, a collection like `/apps` has a set of CRUD methods, and each of them may require different permissions. On the other hand, if an entire collection like `/users` requires the same set of authorisation rules, it is way more convenient to use a middleware.

Reply:
BTW, as discussed, using `authorizer.RequireAdminRole()` instead of `requireAdminRole := authorizer.Require(authz.Role(model.AdminRole))` is definitely preferable, thanks for pointing out this!